wideq/core.py

```diff
@@ -240,166 +324,244 @@ def parse_oauth_callback(url):
-        return cls(data['auth_base'], data['api_root'], data['oauth_root'],
+        return cls(data['auth_base'], data['api_root'], data['api2_root'],
                    data.get('country', DEFAULT_COUNTRY),
                    data.get('language', DEFAULT_LANGUAGE))
 
+    def dump(self):
+        return {
+            "auth_base": self.auth_base,
+            "api_root": self.api_root,
+            "api2_root": self.api2_root,
+            "country": self.country,
+            "language": self.language,
+        }
+
 
 class Auth(object):
-    def __init__(self, gateway, access_token, refresh_token):
+    def __init__(self, gateway, oauth_url, access_token, refresh_token, user_number):
         self.gateway = gateway
+        self.oauth_url = oauth_url
         self.access_token = access_token
         self.refresh_token = refresh_token
+        self.user_number = user_number
 
     @classmethod
     def from_url(cls, gateway, url):
         """Create an authentication using an OAuth callback URL.
         """
+        oauth_url, auth_code, user_number = parse_oauth_callback(url)
+        access_token, refresh_token = login(oauth_url, auth_code)
 
-        access_token, refresh_token = parse_oauth_callback(url)
-        return cls(gateway, access_token, refresh_token)
+        return cls(gateway, oauth_url, access_token, refresh_token, user_number)
 
-    def start_session(self) -> Tuple['Session', List[Dict[str, Any]]]:
+    def start_session(self) -> "Session":
         """Start an API session for the logged-in user. Return the
         Session object and a list of the user's devices.
         """
-
-        session_info = login(self.gateway.api_root, self.access_token,
-                             self.gateway.country, self.gateway.language)
-        session_id = session_info['jsessionId']
-        return Session(self, session_id), get_list(session_info, 'item')
+        return Session(self)
 
     def refresh(self):
         """Refresh the authentication, returning a new Auth object.
         """
 
-        new_access_token = refresh_auth(self.gateway.oauth_root,
-                                        self.refresh_token)
-        return Auth(self.gateway, new_access_token, self.refresh_token)
+        new_access_token = refresh_auth(self.oauth_url, self.refresh_token)
+        return Auth(
+            self.gateway,
+            self.oauth_url,
+            new_access_token,
+            self.refresh_token,
+            self.user_number,
+        )
 
     def serialize(self) -> Dict[str, str]:
         return {
             'access_token': self.access_token,
             'refresh_token': self.refresh_token,
         }
 
+    def dump(self):
+        return {
+            "access_token": self.access_token,
+            "refresh_token": self.refresh_token,
+            "oauth_url": self.oauth_url,
+            "user_number": self.user_number,
+        }
+
+    @classmethod
+    def load(cls, gateway, data):
```

Comment:
For me this always fails, when the new wideq_state.json is written, it only contains "model_info", "gateway" and "auth" as keys.

Reply:
it looks like Auth.dump() in core.py needs to be redirected to Auth.serialize(). One of them should probably be deleted since they're duplicates.